### PR TITLE
docs(contributing): update the coding guideline

### PR DESCRIPTION
**Pull request by Shin-kyoto**
## Description

Updated the coding guideline.

To the reviewers:

- Please check whether there are any inappropriate guidelines included.
- In particular, verify if my thoughts on new/delete/free/raw pointers and throw are correct.
- If there are any other points that should be mentioned, please let me know. I'd like to add anything that can be easily incorporated.
- Please confirm if the structure of the text, the subtitles, and the paragraph divisions are appropriate.
- Regarding the ROS nodes guidelines in the ROS 2 Style Guide, I'm not sure where the link should point to, so I've linked it to class-design.md. If there is a better option, please let me know.
- Since I think that the coding guideline document is still incomplete, I have left the "Under Construction" note. Please let me know if it should be removed.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The Reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by Shin-kyoto** [comment]
@yossytaka
Can you review this PR?

File: docs/contributing/coding-guidelines/index.md
Comment (youtalk) [inline docs/contributing/coding-guidelines/index.md:32]:
```suggestion
- When not using an index, prefer `for (const auto & element : values)`.
```

File: docs/contributing/coding-guidelines/index.md
Comment (youtalk) [inline docs/contributing/coding-guidelines/index.md:44]: The `RCLCPP` is a library name.

```suggestion
- Use `RCLCPP_XXX` (eg. `RCLCPP_INFO`) macros instead of `printf` or `std::cout` for logging.
```

Please modify below too.

File: docs/contributing/coding-guidelines/index.md
Comment (youtalk) [inline docs/contributing/coding-guidelines/index.md:58]: We might add the include naming rule: `include/autoware/XXX` instead of `include/autoware_XXX`.
I'm not sure why this is required though.

File: docs/contributing/coding-guidelines/index.md
Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md]: Thank you so much!!!! :pray:

File: docs/contributing/coding-guidelines/index.md
Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md:50]: I am very grateful for your appropriate comments!! Thank you so much :+1:

File: docs/contributing/coding-guidelines/index.md
Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md:66]: Thank you so much  for your suggestion!!!
I added it in https://github.com/autowarefoundation/autoware-documentation/pull/599/commits/2538ade66d1fb694f0133f73511954bcfb38dfcf

Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md:66]: > I'm not sure why this is required though.

IMO, [this](https://github.com/autowarefoundation/autoware-documentation/blob/f93e489120c1df0f7d4ccb3ebf09a4530be4a56c/docs/contributing/coding-guidelines/ros-nodes/directory-structure.md?plain=1#L211) is the reason.

Comment (youtalk) [reply docs/contributing/coding-guidelines/index.md:66]: Hmm... It's very weak reason.

> Rationale: To avoid verbosity.

**Review by youtalk — APPROVED** [review]
Thank you for your updates.

File: docs/contributing/coding-guidelines/index.md
Comment (sasakisasaki) [inline docs/contributing/coding-guidelines/index.md:43]: Perhaps using `it is` (not `it's`) looks better :+1:

**Review by sasakisasaki — APPROVED** [review]
I have read all contents in the `index.md` and understood this is informative notes. Thank you so much!

File: docs/contributing/coding-guidelines/index.md
Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md]: Thank you so much for your review!! :rocket: 
I fixed in https://github.com/autowarefoundation/autoware-documentation/pull/599/commits/7f812d7b39fa2a8a4b395d5d6d20046ce67f2da8

File: docs/contributing/coding-guidelines/index.md
Comment (mitsudome-r) [inline docs/contributing/coding-guidelines/index.md:27]: Maybe we could also refer to C++ Coding guideline page in Autoware Documentation as well:
https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/

Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md]: Thank you!! I updated in https://github.com/autowarefoundation/autoware-documentation/pull/599/commits/0d2d9c5fa6784d29b45e153ab1b0b17c930293a9

**Comment by mitsudome-r** [comment]
The guideline itself generally looks okay, but most of them seem to be duplicate of the materials in other pages like:
- https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/
- https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/directory-structure/
- https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/parameters/

Could you explain the purpose of this change. Is your intention to add a summary page that require particular attention?

In that case, I'm okay to add Autoware specific guidelines mentioned in this summary page(like adding autoware_ prefix to package names), but I rather keep the details about C++ guidelines as a separate page and just put a link to that page. It would make us harder to keep consistency within our documentation if there are duplicate materials explained in different places.

**Comment by Shin-kyoto** [comment]
> The guideline itself generally looks okay, but most of them seem to be duplicate of the materials in other pages like:
> 
> * https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/
> * https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/directory-structure/
> * https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/parameters/
> 
> Could you explain the purpose of this change. Is your intention to add a summary page that require particular attention?
> 
> In that case, I'm okay to add Autoware specific guidelines mentioned in this summary page(like adding autoware_ prefix to package names), but I rather keep the details about C++ guidelines as a separate page and just put a link to that page. It would make us harder to keep consistency within our documentation if there are duplicate materials explained in different places.

Thank you for your good comment!!
I’m considering making changes based on the following approach. What do you think? @mitsudome-r 

Move the content related to C++ and ROS.
Document Autoware-specific guidelines on this page.

**Comment by TakaHoribe** [comment]
@Shin-kyoto Thank you for your work. Establishing coding guidelines is a great activity as it clarifies the criteria for PRs and reduces review time.

As @mitsudome-r mentioned, the document contains information with different levels of detail and some overlap. For example, the using `.at(i)` instead of `[i]`, and the using the `autoware_` prefix in packages are very different in scope.

It would be helpful to explain why each item is included on this page clearly. For instance, essential points like using the `autoware_` prefix should be clearly stated. On the other hand, for more specific details (like index access or logger usage), it might be better to phrase it as, “Here are some frequent issues raised in reviews or static analysis, referenced from the C++/ROS guidelines".

**Comment by Shin-kyoto** [comment]
> As @mitsudome-r mentioned, the document contains information with different levels of detail and some overlap. For example, the using `.at(i)` instead of `[i]`, and the using the `autoware_` prefix in packages are very different in scope.
> 
> It would be helpful to explain why each item is included on this page clearly. For instance, essential points like using the `autoware_` prefix should be clearly stated. On the other hand, for more specific details (like index access or logger usage), it might be better to phrase it as, “Here are some frequent issues raised in reviews or static analysis, referenced from the C++/ROS guidelines".

Thanks!! :+1: 
I will update document following @mitsudome-r after his answer to https://github.com/autowarefoundation/autoware-documentation/pull/599#issuecomment-2297869052

File: docs/contributing/coding-guidelines/index.md
Comment (TakaHoribe) [inline docs/contributing/coding-guidelines/index.md:20]: suggest to use a specific numbers or reference for better understanding

```suggestion
- Functions that are too complex (low cohesion) should be appropriately split into smaller functions (e.g. less than 40 lines in one function is recommended in the [google style guideline](https://google.github.io/styleguide/cppguide.html#Write_Short_Functions)).
- Try to minimize the use of member variables or global variables that have a large scope.
- Whenever possible, break large pull requests into smaller, manageable PRs (less than 200 is recommended in some research e.g. [here](https://opensource.com/article/18/6/anatomy-perfect-pull-request)).
```

**Comment by mitsudome-r** [comment]
@Shin-kyoto Thanks! 
> Move the content related to C++ and ROS to the separated pages.
Document Autoware-specific guidelines on this page.

I think this approach sounds reasonable to me.

File: docs/contributing/coding-guidelines/index.md
Comment (isamu-takagi) [inline docs/contributing/coding-guidelines/index.md:25]: Since C++ and ROS have their own individual pages, I think it's easy to overlook the information written in the index.

Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md:25]: Thanks! 👍 
Now I'm changing it following this comment!

https://github.com/autowarefoundation/autoware-documentation/pull/599#issuecomment-2297869052

**Comment by Shin-kyoto** [comment]
> The guideline itself generally looks okay, but most of them seem to be duplicate of the materials in other pages like:
> 
> * https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/
> * https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/directory-structure/
> * https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/ros-nodes/parameters/
> 
> Could you explain the purpose of this change. Is your intention to add a summary page that require particular attention?
> 
> In that case, I'm okay to add Autoware specific guidelines mentioned in this summary page(like adding autoware_ prefix to package names), but I rather keep the details about C++ guidelines as a separate page and just put a link to that page. It would make us harder to keep consistency within our documentation if there are duplicate materials explained in different places.

@mitsudome-r @TakaHoribe @isamu-takagi 
I changed in https://github.com/autowarefoundation/autoware-documentation/pull/599/commits/289937d37b92d3c87934738029adce2e975ee108

**Comment by Shin-kyoto** [comment]
@mitsudome-r @TakaHoribe 
Can you review this PR, again?

File: docs/contributing/coding-guidelines/index.md
Comment (mitsudome-r) [inline docs/contributing/coding-guidelines/index.md:34]: Maybe it's better to also add the following documentation link to the reference. The documentation is where we clarify the final decision we made and the discussion explains the background of our decisions.

https://autowarefoundation.github.io/autoware-documentation/pr-599/contributing/coding-guidelines/ros-nodes/directory-structure/#package-name

**Review by mitsudome-r — COMMENTED** [review]
I've made a small suggestions about the reference link. Otherwise, the PR looks good to me.

File: docs/contributing/coding-guidelines/index.md
Comment (Shin-kyoto) [reply docs/contributing/coding-guidelines/index.md:35]: Thank you for your very good feedback!! I updated in https://github.com/autowarefoundation/autoware-documentation/commit/eeee09d06bdd1ce88232bd47e8098b090bdc4bfc

**Review by yukkysaito — APPROVED** [review]

**Review by mitsudome-r — APPROVED** [review]

**Comment by Shin-kyoto** [comment]
Thank you for your review!!
Contribution guideline is very important and it is necessary to review it strictly. Now, I confirmed that
- I got the 5 LGTMs from awf members
- I got the LGTM from maintainer @mitsudome-r
,so I will squash and merge it.

**Pull request merged**
